posthog/clickhouse/cluster.py

```diff
@@ -318,6 +322,30 @@ def __call__(self, client: Client):
         return client.execute(self.query, self.parameters)
 
 
+@dataclass
+class RetryPolicy(Generic[T]):
+    callable: Callable[[Client], T]
+    max_attempts: int
+    delay: float
+    exceptions: tuple[type[Exception], ...] = (Exception,)
+
+    def __call__(self, client: Client) -> T:
+        counter = itertools.count(1)
+        while (attempt := next(counter)) <= self.max_attempts:
+            try:
+                return self.callable(client)
+            except Exception as e:
+                if isinstance(e, self.exceptions) and attempt < self.max_attempts:
+                    logger.warning(
+                        "Failed to invoke %r (attempt #%s), retrying in %s...", self.callable, attempt, self.delay
+                    )
```

Comment:
I'd like to figure out if it's possible to get these logs in the Dagster UI without passing it down the stack everywhere but that seems reasonable to do as a follow up